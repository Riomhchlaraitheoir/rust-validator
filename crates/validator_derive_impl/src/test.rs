use quote::quote;
use syn::DeriveInput;
use crate::Input;

#[test]
fn struct_validator() {
    let input = quote! {
        struct SignupData {
            #[validator(email)]
            mail: String,
            #[validator(url)]
            site: String,
            #[validator(length(min = 1))]
            first_name: String,
            #[validator(range(18..))]
            age: u8,
            #[validator(elements)]
            dogs: Vec<Dog>
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}

#[test]
fn tuple_validator() {
    let input = quote! {
        struct SignupData(
            #[validator(email)]
            String,
            #[validator(url)]
            String,
            #[validator(length(min = 1))]
            String,
        );
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}
#[test]
fn enum_validator() {
    let input = quote! {
        enum Request {
            Signup {
                #[validator(email)]
                mail: String,
                #[validator(url)]
                site: String,
                #[validator(length(min = 1))]
                first_name: String,
            },
            Login(#[validator(email)] String, #[validator(length(min = 8, max = 64))] String),
            Logout
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}


#[test]
fn list_validator() {
    let input = quote! {
        struct HasList {
            #[validator(elements)]
            list: Vec<Element>
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}


