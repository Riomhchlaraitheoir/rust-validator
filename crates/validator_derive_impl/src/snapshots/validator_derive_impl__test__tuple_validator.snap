---
source: crates/validator_derive_impl/src/test.rs
expression: formatted
---
#[derive(Debug)]
struct SignupDataValidationErrors(
    Option<::validator::InvalidEmailError>,
    Option<::validator::InvalidUrlError>,
    Option<::validator::InvalidLengthError>,
);
struct SignupDataValidator(
    ::validator::EmailValidator,
    ::validator::UrlValidator,
    ::validator::LengthValidator,
);
impl ::validator::Validator<SignupData> for SignupDataValidator {
    type Error = SignupDataValidationErrors;
    fn validate(
        &self,
        SignupData(value0, value1, value2): &SignupData,
    ) -> Result<(), Self::Error> {
        let mut _valid = true;
        let validator = self;
        let error = SignupDataValidationErrors {
            0: {
                match validator.0.validate(value0) {
                    Ok(()) => None,
                    Err(error) => {
                        _valid = false;
                        Some(error)
                    }
                }
            },
            1: {
                match validator.1.validate(value1) {
                    Ok(()) => None,
                    Err(error) => {
                        _valid = false;
                        Some(error)
                    }
                }
            },
            2: {
                match validator.2.validate(value2) {
                    Ok(()) => None,
                    Err(error) => {
                        _valid = false;
                        Some(error)
                    }
                }
            },
        };
        if _valid { Ok(()) } else { Err(error) }
    }
}
impl ::validator::Validate for SignupData {
    type Validator = SignupDataValidator;
    fn validator() -> Self::Validator {
        SignupDataValidator {
            0: ::validator::EmailValidator,
            1: ::validator::UrlValidator,
            2: ::validator::LengthValidator::new(Some(1usize), None),
        }
    }
}
